Source/WebKit/UIProcess/API/glib/WebKitImage.cpp

```diff
@@ -0,0 +1,477 @@
+
+    auto* image = WEBKIT_IMAGE(icon);
+
+    Hasher hasher;
+
+    auto* bytes = webkit_image_as_rgba_bytes(image);
+    gsize size = 0;
+    const guint8* data = static_cast<const guint8*>(g_bytes_get_data(bytes, &size));
+    gsize rowBytes = image->priv->width * static_cast<uint8_t>(ImageFormatBPP::RGBA8);
+
+    ASSERT(size >= image->priv->stride * (image->priv->height - 1) + rowBytes);
+
+    for (int y = 0; y < image->priv->height; ++y) {
+        auto span = std::span<const uint8_t>(data + y * image->priv->stride, rowBytes);
+        WTF::add(hasher, span);
+    }
+
+    WTF::add(hasher, image->priv->width, image->priv->height);
+    return hasher.hash();
+}
+
+gboolean webkit_image_equal(GIcon* icon1, GIcon* icon2)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(icon1), false);
+
+    if (!WEBKIT_IS_IMAGE(icon2))
+        return false;
+
+    auto* image1 = WEBKIT_IMAGE(icon1);
+    auto* image2 = WEBKIT_IMAGE(icon2);
+
+    if (image1->priv->width != image2->priv->width
+        || image1->priv->height != image2->priv->height)
+        return false;
+
+    auto* bytes1 = webkit_image_as_rgba_bytes(image1);
+    auto* bytes2 = webkit_image_as_rgba_bytes(image2);
+
+    gsize size1 = 0;
+    gsize size2 = 0;
+    const guint8* data1 = static_cast<const guint8*>(g_bytes_get_data(bytes1, &size1));
+    const guint8* data2 = static_cast<const guint8*>(g_bytes_get_data(bytes2, &size2));
+
+    gsize rowBytes = image1->priv->width * static_cast<uint8_t>(ImageFormatBPP::RGBA8);
+
+    ASSERT(size1 >= image1->priv->stride * (image1->priv->height - 1) + rowBytes);
+    ASSERT(size2 >= image2->priv->stride * (image1->priv->height - 1) + rowBytes);
+
+    for (int y = 0; y < image1->priv->height; ++y) {
+        auto span1 = std::span<const uint8_t>(data1 + y * image1->priv->stride, rowBytes);
+        auto span2 = std::span<const uint8_t>(data2 + y * image2->priv->stride, rowBytes);
+
+        if (!equalSpans(span1, span2))
+            return false;
+    }
+
+    return true;
+}
+
+GInputStream* webkit_image_load(GLoadableIcon* icon, int size, char** type, GCancellable* cancellable, GError** error)
+{
+    g_return_val_if_fail(WEBKIT_IS_IMAGE(icon), nullptr);
+
+    auto* image = WEBKIT_IMAGE(icon);
+    auto* bytes = webkit_image_as_rgba_bytes(image);
+    if (!bytes) {
+        g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_FAILED, "Failed to create image bytes");
+        return nullptr;
+    }
+
+#if USE(SKIA)
+    gsize dataSize = 0;
+    auto* data = g_bytes_get_data(bytes,  &dataSize);
+    sk_sp<SkData> skData = SkData::MakeWithoutCopy(data, dataSize);
+    SkImageInfo info = SkImageInfo::Make(
+        image->priv->width,
+        image->priv->height,
+        kRGBA_8888_SkColorType,
+        kUnpremul_SkAlphaType
+    );
+
+    sk_sp<SkImage> skImage = SkImages::RasterFromData(info, skData, image->priv->stride);
+
+    if (!skImage) {
+        g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_FAILED, "Failed to create SkImage");
```

Comment:
I suppose we do not want to let client code using the API know that we are using Skia here.


```suggestion
        g_set_error_literal(error, G_IO_ERROR, G_IO_ERROR_FAILED, "Failed to encode PNG");
```